### PR TITLE
Feature/pre auth home page

### LINKED ISSUE #72: Customize Auth Guard Redirect Route
#### Describe the desired behavior
Users have no ability to customize the redirect route whenever a block navigation occurs.  Add a config variable so adopters can change this default behavior.  

#### Describe the current behavior
Auth guard only redirects to the '/' root url.

#### Is this request related to a current issue?
no

### PR DESCRIPTION
**Pull request by emclaug2**
Fixes #72 

#### Changes proposed in this Pull Request:
- I added a new prop that allows uses to specify the redirect route whenever the PxbAuthGuard blocks navigation.

-  I made a POC pre-authenticated landing page which is deployed here:  https://pxblue-temp.web.app
- In the code, I've kept the default behavior so that the root url will redirect you to the landing page, not the pre-auth page.

### CODE REVIEW

**Review by joebochill — COMMENTED** [review]
I'm a little confused (probably because I am an angular noob).

How would I configure the router to point any other route back to the route page? With the current configuration, if i enter a route that doesn't exist, I get redirected to the empty route and I'm presented with a blank white screen.

If I attempt to make the empty route the PreAuthComponent, I get stuck in an infinite loop.

I tried adding
```
{ path: '**', redirectTo: '', pathMatch: 'full'},
```
, which works with the config that you left in there, but I get the same infinite loop if I try to make the empty route the pre-auth component.

Basically, I'm trying to make the 'welcome' page the default empty route (/) and I want the auth workflow to redirect me to the empty route after logging in.

**Comment by emclaug2** [comment]
> How would I configure the router to point any other route back to the route page? 
Do you mean the pre-auth page or the auth page here? 

Try something like this:

```
const routes: Routes = [
    { path: '', redirectTo: 'pre-auth', pathMatch: 'full' },
//    { path: '', redirectTo: AUTH_ROUTES.AUTH_WORKFLOW, pathMatch: 'full' },
    { path: AUTH_ROUTES.AUTH_WORKFLOW, component: AuthComponent, children: authWorkflowRoutes },
    {
        path: '',
        canActivate: [PxbAuthGuard],
        children: [
            { path: 'home', component: HomeComponent },
            { path: 'dashboard', component: DashboardComponent },
        ],
    },
    { path: 'pre-auth', component: PreAuthComponent, pathMatch: 'full' },
    { path: '**', redirectTo: '', pathMatch: 'full'},
];
```

The ** path is a catch-all if the previous routes were not used.

**Comment by emclaug2** [comment]
The `authGuardRedirectRoute` isn't what you're thinking it is;  this is the route a user is sent to if they attempt to go to a route that is blocked due to them not being authenticated.

What you're looking for is setting the AUTH_ROUTES.ON_AUTHENTICATED route; this is the route that a user will be directed to when authenticated. 

To use this prop with 0 side-effects, in the app.component.ts make sure to comment out 

```
this.pxbSecurityService.inferOnAuthenticatedRoute('home');
```

This function will parse the URL for you to infer the auth route to use after login.

**Comment by emclaug2** [comment]
Im still thinking about why you've ended up in an infinite loop though.

**Comment by emclaug2** [comment]
This took a little trial and error but this config should do the following:

Load app with `/sample` (route does not exist).

1. Land on pre-auth.
2. Clicking auth-guard button blocks navigation, redirects to same page.
3. Clicking login goes login.
4. Logging in takes you back to pre-auth, since sample doesn't exist.  
> This behavior is weird, not sure what you want here.  We can direct them to /home if the page doesn't exist.


```
AUTH_ROUTES.ON_AUTHENTICATED = 'home';

const authWorkflowRoutes = getAuthSubRoutes();
const routes: Routes = [
    { path: '', redirectTo: 'pre-auth', pathMatch: 'full' },
    { path: 'pre-auth', component: PreAuthComponent, pathMatch: 'full' },
//    { path: '', redirectTo: AUTH_ROUTES.AUTH_WORKFLOW, pathMatch: 'full' },
    { path: AUTH_ROUTES.AUTH_WORKFLOW, component: AuthComponent, children: authWorkflowRoutes },
    {
        path: '',
        canActivate: [PxbAuthGuard],
        children: [
            { path: 'home', component: HomeComponent },
            { path: 'dashboard', component: DashboardComponent },
        ],
    },
    { path: '**', redirectTo: 'pre-auth', pathMatch: 'full'},
];
```


```
this.pxbAuthConfig.authGuardRedirectRoute = 'pre-auth';
```

**Comment by joebochill** [comment]
Okay, so your suggestion about the route config worked. This does exactly what I want:
```tsx
const routes: Routes = [
    { path: '', component: PreAuthComponent, pathMatch: 'full' },
    { path: AUTH_ROUTES.AUTH_WORKFLOW, component: AuthComponent, children: authWorkflowRoutes },
    {
        path: '',
        canActivate: [PxbAuthGuard],
        children: [
            { path: 'home', component: HomeComponent },
            { path: 'dashboard', component: DashboardComponent },
        ],
    },
    
    { path: '**', redirectTo: '', pathMatch: 'full'},
];
```
So that's cool.

Still don't understand what `authGuardRedirectRoute` does. 

If I set it to some value and comment out the route inferring line:
```tsx
        this.pxbAuthConfig.authGuardRedirectRoute = 'testauth';
        // this.pxbSecurityService.inferOnAuthenticatedRoute('home');
```

then I get the following:
1.  Go to `localhost:4200/sample`
2. Get redirected to welcome screen `localhost:4200/`
3. Press button to login, get redirected to `localhost:4200/auth/login`
4. Enter credentials, get redirected to `localhost:4200`
5. Now I can't go to the `/home` page...it just redirects to `/`

So two questions...why can't I get to the home page and where is 'testauth' supposed to come into play?

**Comment by emclaug2** [comment]
> ```tsx
>         this.pxbAuthConfig.authGuardRedirectRoute = 'testauth';
>         // this.pxbSecurityService.inferOnAuthenticatedRoute('home');
> ```
> 
> then I get the following:
> 
> 1. Go to `localhost:4200/sample`
> 2. Get redirected to welcome screen `localhost:4200/`
> 3. Press button to login, get redirected to `localhost:4200/auth/login`
> 4. Enter credentials, get redirected to `localhost:4200`
> 5. Now I can't go to the `/home` page...it just redirects to `/`
> 
> So two questions...why can't I get to the home page and where is 'testauth' supposed to come into play?

`authGuardRedirectRoute` intercepts and redirects unauthed route changes.   If try to go to 'home' without being logged in, you'll get redirected to 'testauth'.    Your app.routing config will redirect this non-existent path back to 'pre-auth'.

Do you have your ON_AUTHENTICATED route set in app.routing.ts? 

AUTH_ROUTES.ON_AUTHENTICATED = 'home';

**Comment by joebochill** [comment]
> `authGuardRedirectRoute` intercepts and redirects unauthed route changes. If try to go to 'home' without being logged in, you'll get redirected to 'testauth'. Your app.routing config will redirect this non-existent path back to 'pre-auth'.

I did not get redirected to the pre-auth page...i got redirected to the login screen. 
And once I was logged in (i.e., I am authenticated at this point) I can see the dashboard page, but not the home page.

> Do you have your ON_AUTHENTICATED route set in app.routing.ts?
> AUTH_ROUTES.ON_AUTHENTICATED = 'home';

No, I did not make changes to that. I don't want to go to home after logging in.


This is the scenario I'm trying to explore:

End user has a product hosted at mydomain.com. They want mydomain.com to hold their main dashboard page (not mydomain.com/home). If they try to go to mydomain.com without logging in, they are forced to login and then get sent to mydomain.com. If they try to go to mydomain.com/page2, they are forced to login and after logging in, they get sent to mydomain.com/page2.

I was trying to think of how to show the welcome page at mydomain.com too, but I don't think that's really possible because the route would have to be simultaneously public and private. So I think i can live with that living at a separate place like mydomain.com/welcome (unless there is a way for the user to set it up as a public route and then conditionally render the app content or the welcome page based on whether the user is authenticated or not).

Am I going down a rabbit hole that I shouldn't be. @daileytj does this make sense for you?

**Comment by joebochill** [comment]
Does the `path` here need to match the `authGuardRedirectRoute` instead of '':
image
?

**Comment by emclaug2** [comment]
@joebochill  I'll update the example with a proper config so example meets expectations.  This should be possible currently.

**Comment by emclaug2** [comment]
This is a lot easier now that I know what you're looking for.  

All this time, you should have been using `pxbSecurityService.inferOnAuthenticatedRoute` because you want the auth module to know which route the user originally intended to navigate to. 

Don't use any routes that don't exist. 

```
        this.pxbAuthConfig.authGuardRedirectRoute = 'pre-auth';
        this.pxbSecurityService.inferOnAuthenticatedRoute('');
```

```

    { path: 'pre-auth', component: PreAuthComponent, pathMatch: 'full' },
    { path: '', component: HomeComponent, pathMatch: 'full' }
    { path: AUTH_ROUTES.AUTH_WORKFLOW, component: AuthComponent, children: authWorkflowRoutes },
    {
        path: '',
        canActivate: [PxbAuthGuard],
        children: [
            { path: '', component: HomeComponent },
            { path: 'home', component: HomeComponent },
            { path: 'dashboard', component: DashboardComponent },
        ],
    },
    { path: '**', redirectTo: 'home', pathMatch: 'full'},
```

Hosted here: https://pxblue-temp.web.app/

If you try to go to a page that does not exist, you will eventually end up at home back.

**Comment by joebochill** [comment]
Cool. I will take a look at this again tomorrow. It might be worth fleshing out the docs a bit (unless I’m just really daft).

**Comment by emclaug2** [comment]
> Cool. I will take a look at this again tomorrow. It might be worth fleshing out the docs a bit (unless I’m just really daft).

I think there's some confusion on both sides; it might be beneficial to doc something up, but it's difficult because the docs would be be more geared towards general routing in angular.

**Comment by emclaug2** [comment]
Fixes #75

File: login-workflow/docs/existing-project-integration.md [hunk 1]
```diff
@@ -44,6 +44,8 @@ In your `app.routing.ts` config, add the auth-specific routes. `authSubRoutes` i
 The configuration below has the base URL redirect to the login screen. 
 All routes that require authentication can be protected using the `PxbAuthGuard`.  This guard will prevent routes from being accessed from unauthenticated users.  
 
+> By default, the `PxbAuthGuard` will redirect an unauthenticated user to the empty route; the redirect url can be configured in the `PxbAuthConfig` service.
```
Comment (joebochill) [inline login-workflow/docs/existing-project-integration.md:47]:
```suggestion
> By default, the `PxbAuthGuard` will redirect an unauthenticated user to the `AUTH_ROUTES.AUTH_WORKFLOW` route. If you would like to redirect the users to a different screen, the destination url can be configured in the `PxbAuthConfig` service with the `authGuardRedirectRoute` property.
```

**Review by joebochill — APPROVED** [review]
